corehq/apps/reports/standard/deployments.py

```diff
@@ -272,22 +272,29 @@ def get_bulk_ancestors(self, location_ids):
         :param kwargs: extra parameters
         :return: dict
         """
-        where = Q(domain=self.domain, location_id__in=location_ids)
-        location_ancestors = SQLLocation.objects.get_ancestors(where)
-        location_by_id = {location.location_id: location for location in location_ancestors}
-        location_by_pk = {location.id: location for location in location_ancestors}
-
-        grouped_location = {}
-        for location_id in location_ids:
-
-            location_parents = []
-            current_location = location_by_id[location_id].id if location_id in location_by_id else None
+        # list of ancestors of a location
+        ancestors_by_location = {}
+
+        def get_ancestors(loc):
+            if loc.pk in ancestors_by_location:
+                return ancestors_by_location[loc.pk]
+            all_ancestors = []
+            if loc.parent:
+                all_ancestors.extend(loc.parent)
+                if loc.parent_id in ancestors_by_location:
+                    parent_ancestors = ancestors_by_location[loc.parent_id]
+                else:
+                    parent_ancestors = get_ancestors(loc.parent)
+                all_ancestors.extend(parent_ancestors)
 
-            while current_location is not None:
-                location_parents.append(location_by_pk[current_location])
-                current_location = location_by_pk[current_location].parent_id
+            # note down ancestors for further searches
+            ancestors_by_location[loc.pk] = all_ancestors
+            return all_ancestors
 
-            grouped_location[location_id] = location_parents
+        grouped_location = {}
+        for location in SQLLocation.objects.select_related('parent').filter(
+                domain=self.domain, location_id__in=location_ids):
+            grouped_location[location.location_id] = get_ancestors(location)
```

Comment:
isn't `grouped_location` identical to `ancestors_by_location`?  Why not just populate and return that?

Reply:
`ancestors_by_location` would be ancestors for any location, which means the locations asked for or any of their individual ancestors as well. It is a local var mainly for record keeping to avoid fetching them again.
We could then either pop or fetch only the locations we need in the end, which might get a little more complicated i guess.

Reply:
Ah, you're right, I didn't realize that.  I don't think it makes a difference in how the return value is used, though - since it's just a cache.  The `get_location_columns` business iterates through the whole cache, but that also shouldn't matter, since it's ancestors only.  This builds up a second dictionary, which will use more memory in the process, but then `ancestors_by_location` will be freed up after this method completes, and that's marginally larger.  I don't feel too strongly about it either way.